File: jar/src/main/kotlin/jline/api/mam/Dmap_moment.kt
/**
 * Moment computation for discrete-time MAPs.
 *
 * @since LINE 3.0
 */
package jline.api.mam

import jline.util.matrix.Matrix
import jline.util.matrix.MatrixCell

/**
 * Computes the k-th raw moment of the inter-arrival time of a discrete MAP.
 *
 *   E[T]   = pi * (I-D0)^{-1} * e
 *   E[T^2] = 2*pi*(I-D0)^{-2}*e - E[T]
 *   E[T^3] = 6*pi*(I-D0)^{-3}*e - 6*pi*(I-D0)^{-2}*e + E[T]
 */
fun dmap_moment(D0: Matrix, D1: Matrix, order: Int): Double {
    val n = D0.numRows
    val I = Matrix.eye(n)
    val ImD0inv = I.add(-1.0, D0).inv()
    val e = Matrix.ones(n, 1)
    val pi = dmap_pie(D0, D1)

    return when (order) {
        1 -> pi.mult(ImD0inv).mult(e).toDouble()
        2 -> {
            val m1 = pi.mult(ImD0inv).mult(e).toDouble()
            2.0 * pi.mult(ImD0inv.mult(ImD0inv)).mult(e).toDouble() - m1
        }
        3 -> {
            val m1 = pi.mult(ImD0inv).mult(e).toDouble()
            val ImD0inv2 = ImD0inv.mult(ImD0inv)
            6.0 * pi.mult(ImD0inv2.mult(ImD0inv)).mult(e).toDouble() - 6.0 * pi.mult(ImD0inv2).mult(e).toDouble() + m1
        }
        else -> throw IllegalArgumentException("Moments of order > 3 not implemented for DMAP")
    }
}

fun dmap_moment(DMAP: MatrixCell, order: Int): Double = dmap_moment(DMAP[0], DMAP[1], order)

File: matlab/lib/kpctoolbox/dmap/dmap_moment.m
function MOMENTS=dmap_moment(DMAP,ORDERS)
% MOMENTS=dmap_moment(DMAP,ORDERS) - Compute factorial moments of
% inter-arrival times for a discrete-time MAP
%
%  Input:
%  DMAP: a D-MAP in the form of {D0,D1}
%  ORDERS: set of moment orders (1=>E[X], 2=>E[X^2], ...)
%
%  Output:
%  MOMENTS: moments returned in the same order of ORDERS

D0=DMAP{1};
D1=DMAP{2};
N=size(D0,1);
e=ones(N,1);
P=inv(eye(N)-D0)*D1;
al=dtmc_solve(P);
for t=1:length(ORDERS)
    i=ORDERS(t);
    if isnan(D0)
        MOMENTS(t)=NaN;
    else
        A=(inv(eye(N)-D0)^i);
        MOMENTS(t)=factorial(i)*al*A*e;
    end
end
end
